### PR TITLE
Update to python3

### PR DESCRIPTION
**Pull request by sschiessl-bcp**

### CODE REVIEW

File: api/explorer.py [hunk 1]
```diff
@@ -5,10 +5,8 @@
 from services.bitshares_websocket_client import client as bitshares_ws_client
 from services.bitshares_elasticsearch_client import client as bitshares_es_client
 from services.cache import cache
-import es_wrapper
+from . import es_wrapper
```
Comment (oxarbitrage) [inline api/explorer.py:8]: why change here?

File: api/explorer.py [hunk 2]
```diff
@@ -5,10 +5,8 @@
...
 import config
-from app import limiter
```
Comment (oxarbitrage) [inline api/explorer.py:10]: why remove limiter ?

File: api/nomics.py [hunk 3]
```diff
@@ -5,8 +5,8 @@
 from services.bitshares_websocket_client import client as bitshares_ws_client
 from services.bitshares_elasticsearch_client import client as bitshares_es_client
 from services.cache import cache
-import api.explorer
-import es_wrapper
+from . import explorer
+from . import es_wrapper
```
Comment (oxarbitrage) [inline api/nomics.py:9]: why this is changing ?

File: api/utils.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+from functools import wraps
```
Comment (oxarbitrage) [inline api/utils.py:1]: i don't see this file used anywhere nor the functions inside it.

**Review by oxarbitrage — COMMENTED** [review]
Basically the only change needed to use python 3.7 is to change the `virtualenv`?

Did you tested this changes in your end ?

File: api/explorer.py (hunk 1 above)
Comment (sschiessl-bcp) [reply api/explorer.py:8]: python3 needs strict relative imports

File: api/explorer.py (hunk 2 above)
Comment (sschiessl-bcp) [reply api/explorer.py:10]: I couldn't get the import to work, will need to revisit .. this should not have been in this PR.

File: api/nomics.py (hunk 3 above)
Comment (sschiessl-bcp) [reply api/nomics.py:9]: python3 needs strict relative imports

File: api/utils.py (hunk 4 above)
Comment (sschiessl-bcp) [reply api/utils.py:1]: Ah .. yes. It is currently used in the voting-stats part (which is not part of this PR). 

Essentially, all queries that go to ES can be wrapped with this for a stringent error message.

**Comment by sschiessl-bcp** [comment]
> Basically the only change needed to use python 3.7 is to change the `virtualenv`?
> 
> Did you tested this changes in your end ?

Yes, this is the version that is live @ http://88.198.69.93:9011/apidocs/

File: api/explorer.py (hunk 2 above)
Comment (oxarbitrage) [reply api/explorer.py:10]: You probably need to run `pip install -r requirements/production.pip` again.

Flask-limiter was added there. https://github.com/bitshares/bitshares-explorer-api/blob/master/requirements/production.pip#L5

If that works for you lets keep it.

File: api/utils.py (hunk 4 above)
Comment (oxarbitrage) [reply api/utils.py:1]: I think we should leave it out from this PR. Will be great if you can submit another one with this and the voting-stats when ready :)

**Comment by oxarbitrage** [comment]
That is great, i always believed more work was involved to do the change so i had been postponing it.
This is important as 2.7 supporting is ending jan 2020.

File: api/utils.py (hunk 4 above)
Comment (sschiessl-bcp) [reply api/utils.py:1]: I all add it to another call :) It's useful to identify properly what is supported (IMHO)

File: api/explorer.py (hunk 2 above)
Comment (sschiessl-bcp) [reply api/explorer.py:10]: python3 does not allow import cyclic imports like that anymore, so we will need to restructure the app a bit to allow that.

Two ways:
a) I do the restructing within this PR
b) I remove limiter for now, and then do a restructure-only PR

I would vote for b), what do you think?

File: api/utils.py (hunk 4 above)
Comment (sschiessl-bcp) [reply api/utils.py:1]: Added now response checking as well!

File: api/explorer.py (hunk 2 above)
Comment (oxarbitrage) [reply api/explorer.py:10]: Ok, lets do that.

**Review by oxarbitrage — APPROVED** [review]

**Pull request merged**